In [1]:
import pandas as pd

df = pd.read_csv("olist_order_items_dataset.csv")
df.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


Ici j'affiche le résumé de la table, nombre de lignes
nombre de colonnes
nom des colonnes
type des données (texte, nombre…); ça me permet de voir à quoi ressemble mon fichier avant de faire des calculs.


In [3]:
df.head(10)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
5,00048cc3ae777c65dbb7d2a0634bc1ea,1,ef92defde845ab8450f9d70c526ef70f,6426d21aca402a131fc0a5d0960a3c90,2017-05-23 03:55:27,21.90,12.69
6,00054e8431b9d7675808bcb819fb4a32,1,8d4f2bb7e93e6710a28f34fa83ee7d28,7040e82f899a04d1b434b795a43b4617,2017-12-14 12:10:31,19.90,11.85
7,000576fe39319847cbb9d288c5617fa6,1,557d850972a7d6f792fd18ae1400d9b6,5996cddab893a4652a15592fb58ab8db,2018-07-10 12:30:45,810.00,70.75
8,0005a1a1728c9d785b8e2b08b904576c,1,310ae3c140ff94b03219ad0adc3c778f,a416b6a846a11724393025641d4edd5e,2018-03-26 18:31:29,145.95,11.65
9,0005f50442cb953dcd1d21e1fb923495,1,4535b0e1091c278dfd193e5a1d63b39f,ba143b05f0110f0dc71ad71b4466ce92,2018-07-06 14:10:56,53.99,11.40


J'affiche les 10 premières lignes de la table afin de voir des exemples réels de données.

In [4]:
df['order_id'].nunique()

98666

Je vérifie le nombre de commandes uniques, parce qu'une commande peut apparaitre plusieurs fois (commande ayant plsusieurs produits)

In [5]:
df.shape

(112650, 7)

J'affiche le nombre de lignes et de colonnes pour avoir une idée de la taille du dataset

In [6]:
df['order_item_id'].nunique()

21

Je compte le nombre de valeurs différentes présentes dans la colonne order_item_id, car lorsque j'ai affiché les 10 premières lignes du dataset j'ai remarqué que le chiffre 1 se répétait dans celle-ci

In [7]:
df['order_item_id'].value_counts().head(10)

,count
order_item_id,
1,98666
2,9803
3,2287
4,965
5,460
6,256
7,58
8,36
9,28


Ce résultat montre que certaines commandes ont plusieurs produits

In [8]:
df[df['order_item_id'] > 1].head(10)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
14,0008288aa423d2a3f00fcb17cd7d8719,2,368c6c730842d78016ad823897a372db,1f50f920176fa81dab994f9023523100,2018-02-21 02:55:52,49.90,13.37
33,00143d0f86d6fbd9f9b38ab440ac16f5,2,e95ee6822b66ac6058e2e4aff656071a,a17f621c590ea0fab3d5d883e1630ec6,2017-10-20 16:07:52,21.33,15.10
34,00143d0f86d6fbd9f9b38ab440ac16f5,3,e95ee6822b66ac6058e2e4aff656071a,a17f621c590ea0fab3d5d883e1630ec6,2017-10-20 16:07:52,21.33,15.10
43,001ab0a7578dd66cd4b0a71f5b6e1e41,2,0b0172eb0fd18479d29c3bc122c058c2,5656537e588803a555b8eb41f07a944b,2018-01-04 02:33:42,24.89,17.63
44,001ab0a7578dd66cd4b0a71f5b6e1e41,3,0b0172eb0fd18479d29c3bc122c058c2,5656537e588803a555b8eb41f07a944b,2018-01-04 02:33:42,24.89,17.63
49,001d8f0e34a38c37f7dba2a37d4eba8b,2,e67307ff0f15ade43fcb6e670be7a74c,f4aba7c0bca51484c30ab7bdc34bcdd1,2017-05-18 17:35:11,18.99,7.78
76,002c9def9c9b951b1bec6d50753c9891,2,2d9ff06c8870a518f5f6909774e140fb,00720abe85ba0859807595bbf045a33b,2018-05-21 17:32:03,78.00,8.90
81,002f98c0f7efd42638ed6100ca699b42,2,880be32f4db1d9f6e2bec38fb6ac23ab,fa40cc5b934574b62717c68f3d678b6d,2017-08-10 09:30:15,44.90,7.16
88,003324c70b19a16798817b2b3640e721,2,2b939dc9b176d7fa21594d588815d4a4,dbc22125167c298ef99da25668e1011f,2017-05-24 19:22:34,102.90,14.45
92,00337fe25a3780b3424d9ad7c5a4b35e,2,13944d17b257432717fd260e69853140,cfb1a033743668a192316f3c6d1d2671,2017-09-29 17:50:16,59.90,9.94


In [9]:
df.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')

J'affiche la liste des colonnes de la table order_items afin d'identifier les informations disponibles et de déterminer celles qui seront utiles pour construire la future table de ventes.

Après avoir identifiée les colonnes de la table order_items, j'ai distingué: les mesures à analyser (price, freight_value); et les identifiants permettant de relier les données (order_id, product_id, et seller_id)

In [10]:
fact_sales = df[['order_id',
                 'product_id',
                 'seller_id',
                 'price',
                 'freight_value']]

J'ai identifié la table order_items comme la base de ma future table de faits car chaque ligne représente un produit vendu dans une commande.

Pour ce faire, j'ai donc conservé les colonnes nécessaires à l'analyse des ventes (price, freight_value); ainsi que les identifiants permettant de relier les données à d'autres tables(order_id, product_id, & seller_id).

In [11]:
fact_sales.head()

,order_id,product_id,seller_id,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,199.90,18.14


J'affiche les premières lignes de ma nouvelle table, pour vérifier que le résultat est conforme à celui que j'attend.

La table de faits nommée fact_sales que j'ai créée regroupe les informations liées aux ventes et contient les mesures à analyser (prix, frais de livraison). Ainsi que les identifiants nécessaires pour relier les ventes aux autres données du système

# Construire notre première dimension

Je vais maintenant construire notre première dimension. Une dimension est une table qui décrit quelque chose. Par exemple, sur un ticket de caisse il est marqué réf. produit et prix, on sait qu'un produit a été vendu. Mais, on ignore le nom du produit, sa catégorie, et ses caratéristiques.

Pour le savoir, il nous faudra une table, dim_product. Avant de créer la dimension produit, je crée en amont la variable produit.

In [12]:
products = pd.read_csv("olist_products_dataset.csv")
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


J'ai chargé le fichier olist_products_dataset.csv qui contient les informations descriptives des produits. Car ces données serviront à construire la dimension produit du data warehouse.

In [13]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


J'explore la structure du dataset afin d'identifier les attributs disponibles pour décrire les produits.

In [14]:
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')

J'affiche les colonnes du dataset afin d'identifier les attributs disponibles pour construire la dimension produit

In [15]:
dim_product = products[['product_id',
                        'product_category_name',
                        'product_photos_qty',
                        'product_weight_g',
                        'product_length_cm',
                        'product_height_cm',
                        'product_width_cm']]

La dimension dim_product est désormais créée, en ayant sélectionnée uniquement les attributs jugés utiles aux futures analyses:catégorie, nombre de photos, poids et dimensions du produit. J'ai conservé également l'identifiant produit afin de permettre les relations avec la table de faits.

In [16]:
dim_product.head()

,product_id,product_category_name,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,4.0,625.0,20.0,17.0,13.0


Je contrôle toujours le résultat de la transformation afin de vérifier que les colonnes sélectionnées sont correctement présentes dans la dimension produit.

In [17]:
dim_product.shape

(32951, 7)

Je vérifie également la taille de la dimension produit afin de m'assurer que seules les colonnes ont été modifiées et le nombre d'enregistrement reste correct

In [18]:
dim_product.isnull().sum()

,0
product_id,0
product_category_name,610
product_photos_qty,610
product_weight_g,2
product_length_cm,2
product_height_cm,2
product_width_cm,2


Ici j'analyse les valeurs manquantes de la dimension produit afin d'évaluer la qualité des données avant leur intégration dans le data warehouse.

Celle-ci révèle qu'il n'y a aucune valeur manquante pour l'identifiant produit. Cependant, quelques attributs présentent des données absentes, notamment la catégorie du produit et le nombre de photos (610 enregistrements); ainsi que certaines dimensions physiques (2 enregistrements ). A ce stade du projet, ces valeurs sont conservées afin de préserver l'intégralité des données sources.

In [19]:
sellers = pd.read_csv("olist_sellers_dataset.csv")
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


Afin de de construire la dimension vendeur du data warehouse, je charge le fichier des vendeurs. Cette dimension nous permettra de répondre à des questions comme: quel vendeur réalise le plus de ventes , ou quel vender génère le plus de chiffres d'affaires.

In [20]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


J'explore la structure du dataset vendeur afin d'identifier les attributs disponibles pour décrire les vendeurs.

In [21]:
sellers.columns

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='object')

Maintenant, j'affiche les colonnes du dataset, afin d'identifier les attributs pertinents pour la construction de la dimension

In [22]:
dim_seller = sellers[['seller_id',
                      'seller_zip_code_prefix',
                      'seller_city',
                      'seller_state']]
dim_seller.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [23]:
dim_seller.shape

(3095, 4)

La dimension vendeur contient 3 095 enregistrements et 4 attributs descriptifs permettant d'identifier et de localiser les vendeurs.

In [24]:
dim_seller.isnull().sum()

,0
seller_id,0
seller_zip_code_prefix,0
seller_city,0
seller_state,0


Après vérification des valeurs manquantes, on peut observer qu'il n'y a aucune valeur manquante. Les données de localisation des vendeurs sont donc complètes et directement exploitables pour les futures analyse géographiques.

In [25]:
orders = pd.read_csv("olist_orders_dataset.csv")
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


Ici, je charge le fichier des commandes. Celui-ci est très intéressante car elle contient la notion de temps.

In [26]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


Comme avant chaque construction de dimension, j'analyse la structure du dataset afin d'identifier les attributs disponibles pour décrire les commandes.

In [27]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

Ici, j'affiche les colonnes du dataset afin d'identifier les attributs pertinents pour la construction de la dimension

In [28]:
dim_order = orders[['order_id',
                    'customer_id',
                    'order_status',
                    'order_purchase_timestamp',
                    'order_approved_at',
                    'order_delivered_carrier_date',
                    'order_delivered_customer_date',
                    'order_estimated_delivery_date']]
dim_order.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


Après avoir analysé les colonnes du dataset, je conserve l'ensemble des attributs car ils permettent tous de suivre le cycle de vie complet d'une commande, depuis son achat jusqu'à la livraison. Ils offrent également la possibilté d'analyser les délais, les statuts et les performances logistiques.

In [29]:
dim_order.shape

(99441, 8)

In [30]:
dim_order.isnull().sum()

,0
order_id,0
customer_id,0
order_status,0
order_purchase_timestamp,0
order_approved_at,160
order_delivered_carrier_date,1783
order_delivered_customer_date,2965
order_estimated_delivery_date,0


L'analyse des valeurs manquantes montre l'absence des données sur certaines étapes du cycle de vie des commandes. Ces absences concernent principalemnt les dates d'approbation et de livraison. Elles sont problablement liées à des commandes annulées ou non encire finalisées et ne construisent donc pas nécessairement une anomalie.

In [31]:
dim_order['order_status'].value_counts()

,count
order_status,
delivered,96478
shipped,1107
canceled,625
unavailable,609
invoiced,314
processing,301
created,5
approved,2


L'analyse des statuts montre que 96 478 commandes sur 99 441 ont été livrées. Le nombre de valeurs manquantes observé dans la date de livraison client est cohérent avec le nombre de commandes dont le statut n'est pas "delivered". Ces absences semblent donc liées au processus métier plutôt qu'à un problème de qualité des données.

Maintenant que les tables du data warehouse sont construites (fact_sales et les dimensions produits, vendeurs et commandes), je peux passer à une phase d'analyse.

L'objectif est de produire des indicateurs métier simples afin de comprendre la performance commerciale de l'e-commerce

In [32]:
total_ca = fact_sales['price'].sum()
total_ca

np.float64(13591643.700000003)

Le chiffre d'affaires total représente la somme de toutes les ventes enregistrées dans la table de faits. Celui-ci représente la performance globale du e-commerce sur la période couverte par les données

Maintenant que nous avons le chiffres d'affaires total, on peut analyser sa répartition par catégorie de produit.

L'objectif est d'identifier les catégories les plus contributrices au CA et de mieux comprendre les segments les plus performants du e-commerce.

Pour cela, je dois combiner la table des ventes(fact_sales) avec la dimension produit(dim_product).

In [33]:
df = fact_sales.merge(dim_product, on='product_id', how='left')
df.head()

,order_id,product_id,seller_id,price,freight_value,product_category_name,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,00010242fe8c5a6d1ba2dd792cb16214,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,58.90,13.29,cool_stuff,4.0,650.0,28.0,9.0,14.0
1,00018f77f2f0320c557190d7a144bdd3,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,239.90,19.93,pet_shop,2.0,30000.0,50.0,30.0,40.0
2,000229ec398224ef6ca0657da4fc703e,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,199.00,17.87,moveis_decoracao,2.0,3050.0,33.0,13.0,33.0
3,00024acbcdf0a6daa1e931b038114c75,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,12.99,12.79,perfumaria,1.0,200.0,16.0,10.0,15.0
4,00042b26cf59d7ce69dfabb4e55b4fd9,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,199.90,18.14,ferramentas_jardim,1.0,3750.0,35.0,40.0,30.0


In [34]:
ca_category = df.groupby('product_category_name')['price'].sum().sort_values(ascending=False)
ca_category

,price
product_category_name,
beleza_saude,1258681.34
relogios_presentes,1205005.68
cama_mesa_banho,1036988.68
esporte_lazer,988048.97
informatica_acessorios,911954.32
...,...
flores,1110.04
casa_conforto_2,760.27
cds_dvds_musicais,730.00


Le CA a été agrégé par catégorie de produit afin d'identifier les segments les plus performants.

Cette analyse permet de comprendre quelles catégories génèrent le plus de revenus et peuvent représenter des axes stratégiques pour l'entreprise.

Après avoir analysé le CA par catégorie de produit, on va maintenant identifier les vendeurs les plus performants.

L'objectif étant de comprendre quels vendeurs  génèrent le plus de revenus et contribuent le plus au CA global.

In [35]:
top_sellers = fact_sales.groupby('seller_id')['price'].sum().sort_values(ascending=False)
top_sellers.head(10)

,price
seller_id,
4869f7a5dfa277a7dca6462dcf3b52b2,229472.63
53243585a1d6dc2643021fd1853d8905,222776.05
4a3ca9315b744ce9f8e9374361493884,200472.92
fa1c13f2614d7b5c4749cbc52fecda94,194042.03
7c67e1448b00f6e969d365cea6b010ab,187923.89
7e93a43ef30c4f03f38b393420bc753a,176431.87
da8622b14eb17ae2831f4ac5b9dab84a,160236.57
7a67c85e85bb2ce8582c35f2203ad736,141745.53
1025f0e2d44d7041d6cf58b6550e0bfa,138968.55


Le CA a été agrégé par vendeur afin d'identifier les meilleurs contributeurs aux ventes.
Cette analyse permet de mettre en évidence les vendeurs les plus performants et potentiellment stratégiques pour la marketplace.

A présent, je vais analyser la performance logistique des commandes en étudiant les délais de livraison.
L'objectif étant de comprendre le temps réel entre la commande et la réception par client. Ainsi que d'évaluer si les livraisons respectent les délais estimés.

In [36]:
dim_order['order_purchase_timestamp'] = pd.to_datetime(dim_order['order_purchase_timestamp'])
dim_order['order_delivered_customer_date'] = pd.to_datetime(dim_order['order_delivered_customer_date'])

#je transforme les colonnes en vraies dates, parce que si elles restent telles quelles, python ne pourra pas faire le calcul

In [37]:
dim_order['delivery_time_days'] = (
    dim_order['order_delivered_customer_date'] - dim_order['order_purchase_timestamp']
).dt.days

#je tranforme le résultat en nbre de jours

In [38]:
dim_order['delivery_time_days'].mean() # je calcule le délai moyen de livraison avec la viariable que j'ai créée et stockée

np.float64(12.094085575687217)

In [39]:
avg_delivery = dim_order['delivery_time_days'].mean() # je crée et stocke la variabe avg_delivery si je dois la réutiliser plus tard, l'afficher proprement et/ou l'intégrer dans un rapport

Le délai de livraison correspond au temps écoulé entre la commande et la reception par client.
Le calcul montre un délai moyen d'environ 12 jours.

Cet indicateur permet d'évaluer la performance logistique globale du marketplace.

Maintenant que le délai moyen de livraison a été calculé, on peut analyser la fiabilité des délais annoncés aux clients.

L'objectif ici étant de déterminer si les commandes sont livrées en retard par rapport à la date de livraison estimée, et dans quelle proportion

In [40]:
#je vais convertir les dates, elles sont en str et je ne peux pas calculer Timestamp avec
dim_order['order_delivered_customer_date'] = pd.to_datetime(dim_order['order_delivered_customer_date'])
dim_order['order_estimated_delivery_date'] = pd.to_datetime(dim_order['order_estimated_delivery_date'])

In [41]:
# puis je crée la variable du retard et je fais le calcul entre la date de livraison réelle et celle prévue
dim_order['delay_vs_estimation'] = (
    dim_order['order_delivered_customer_date'] - dim_order['order_estimated_delivery_date']
).dt.days

dim_order['delay_vs_estimation'].head()

,delay_vs_estimation
0,-8.0
1,-6.0
2,-18.0
3,-13.0
4,-10.0


In [42]:
#je calcule maintenant si globalement les livraisons sont en avance ou en retard
dim_order['delay_vs_estimation'].mean()

np.float64(-11.876881296902857)

In [43]:
avg_delay_vs_estimation = dim_order['delay_vs_estimation'].mean()

L'analye des écarts entre la date estimée et la date réelle de livraison permet  donc d'évaluer la fiabilité des délais annoncés aux clients.

Un écart négatif indique qu'une livraison est en avance, tandis qu'un écart positif indique qu'elle est en retard.

Cette métrique est essentielle pour mesurer la performance logistique de la market place.

Après avoir calculé l'écart entre la date estimée et la date réelle de livraison, on va mesuer dès à présent la proportion de commandes livrées en retard.
Cet indicateur permet d'évaluer l'impact réel des retards sur les clients et de compléter l'analyse du délai moyen de livraison.

In [44]:
# j'identifie les commandes en retard
dim_order['delay_vs_estimation'] > 0

,delay_vs_estimation
0,False
1,False
2,False
3,False
4,False
...,...
99436,False
99437,False
99438,False
99439,False


In [45]:
# je calcule le pourcentage
(dim_order['delay_vs_estimation'] > 0 ).mean() *100

np.float64(6.571736004263835)

L'analyse des délais de livraison montre qu'une commande est livrée en moyenne en 12 jours.
De plus, seulement environ 6,6% des commandes ont été livrées après la date estimée.
Ces résultats indiquent une bonne fiabilité logistique. La majorité des commandes étant livrées dans les délais prévus ou en avance.

# Conculison générale

Dans ce projet, j'ai conçu un mini data warehouse à partir du jeu de données e-commerce Olist.

Après l'identification des différentes sources de données, j'ai construit une table de faits regroupant les ventes ainsi que plusieurs dimensions dédiées aux produitx, aux vendeurs et aux commandes.
Cette modélisation a permis de centraliser les informations nécessaires à l'analyse des performances commerciales et logistiques.

Plusieurs indicateurs métier ont ensuite été calculés, notamment le chiffre d'affaires total, le chiffre d'affaires total par catégorie de produit, ainsi que les indicateurs liés aux délais de livraison.

Ce projet m'a permis de mettre en pratique des compétences en péparation de données, modélisation de données, manipulation avec python et analyse métier. Il constitue une approche des problématiques rencontrées dans les environnements data analys et data engineer.